File: brain/agent.py
import anthropic
from config import ANTHROPIC_API_KEY, CLAUDE_MODEL, CLIENTS
from brain.prompts import build_system_prompt
from brain.context import load_client_context
from tools.research import get_research_tools
from tools.campaigns import get_campaign_tools
from tools.inbox import get_inbox_tools
import tools.research as _t_research
import tools.campaigns as _t_campaigns
import tools.inbox as _t_inbox
from rich.console import Console
from rich.markdown import Markdown

console = Console()
client = anthropic.Anthropic(api_key=ANTHROPIC_API_KEY)


def get_all_tools():
    return [
        *get_research_tools(),
        *get_campaign_tools(),
        *get_inbox_tools(),
    ]


def _wire_client(client_name: str) -> None:
    config = next((c for c in CLIENTS if c["name"] == client_name), {})
    _t_research.set_client(config)
    _t_campaigns.set_client(config)
    _t_inbox.set_client(config)


def run_agent(user_message: str, conversation_history: list = None, client_name: str = None):
    if conversation_history is None:
        conversation_history = []

    if client_name:
        _wire_client(client_name)

    context = load_client_context(client_name) if client_name else ""
    system = build_system_prompt(context)

    conversation_history.append({"role": "user", "content": user_message})

    tools = get_all_tools()

    while True:
        response = client.messages.create(
            model=CLAUDE_MODEL,
            max_tokens=4096,
            system=system,
            tools=tools,
            messages=conversation_history,
        )

        assistant_content = response.content
        conversation_history.append({"role": "assistant", "content": assistant_content})

        if response.stop_reason == "end_turn":
            for block in assistant_content:
                if hasattr(block, "text"):
                    console.print(Markdown(block.text))
            break

        if response.stop_reason == "tool_use":
            tool_results = []
            for block in assistant_content:
                if block.type == "tool_use":
                    console.print(f"\n[bold cyan]Using tool:[/bold cyan] {block.name}")
                    result = _execute_tool(block.name, block.input)
                    console.print(f"[dim]{str(result)[:200]}...[/dim]" if len(str(result)) > 200 else f"[dim]{result}[/dim]")
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block.id,
                        "content": str(result),
                    })

            conversation_history.append({"role": "user", "content": tool_results})

    return conversation_history


def _execute_tool(tool_name: str, tool_input: dict):
    from tools.research import execute_research_tool
    from tools.campaigns import execute_campaign_tool
    from tools.inbox import execute_inbox_tool

    for executor in [execute_research_tool, execute_campaign_tool, execute_inbox_tool]:
        result = executor(tool_name, tool_input)
        if result is not None:
            return result

    return {"error": f"Unknown tool: {tool_name}"}

File: mcp/lib/agent.js
// ── agent.js ─────────────────────────────────────────────────────────────────
// Generic delegating-agent runner. Each operator agent = a system prompt + a set
// of tools (each tool: {name, description, input_schema, handler}). The operator
// hands the agent a goal; this runs an Anthropic tool-use loop until the agent
// finishes, then returns its final message plus a transcript of tool calls.
import Anthropic from "@anthropic-ai/sdk";

const MODEL       = process.env.AGENT_MODEL || "claude-sonnet-4-6";
const MAX_TURNS   = Number(process.env.AGENT_MAX_TURNS || 25);
const MAX_TOKENS  = Number(process.env.AGENT_MAX_TOKENS || 4096);

let _anthropic = null;
function client() {
  if (_anthropic) return _anthropic;
  if (!process.env.ANTHROPIC_API_KEY) {
    throw new Error("ANTHROPIC_API_KEY is not set — required to run delegating agents.");
  }
  _anthropic = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });
  return _anthropic;
}

/**
 * @param {object} agent  { name, systemPrompt, tools: [{name, description, input_schema, handler}] }
 * @param {string} goal   Natural-language objective from the operator.
 * @param {object} [opts] { context?: string }
 * @returns {Promise<{ agent, final_text, tool_calls, turns }>}
 */
export async function runAgent(agent, goal, opts = {}) {
  const anthropic = client();
  const toolMap = Object.fromEntries(agent.tools.map(t => [t.name, t]));
  const apiTools = agent.tools.map(t => ({
    name: t.name,
    description: t.description,
    input_schema: t.input_schema,
  }));

  const system = opts.context
    ? `${agent.systemPrompt}\n\n## Operator context\n${opts.context}`
    : agent.systemPrompt;

  const messages = [{ role: "user", content: goal }];
  const toolCalls = [];
  let turns = 0;

  while (turns < MAX_TURNS) {
    turns++;
    const resp = await anthropic.messages.create({
      model: MODEL,
      max_tokens: MAX_TOKENS,
      system,
      tools: apiTools,
      messages,
    });

    messages.push({ role: "assistant", content: resp.content });

    if (resp.stop_reason !== "tool_use") {
      const finalText = resp.content.filter(b => b.type === "text").map(b => b.text).join("\n").trim();
      return { agent: agent.name, final_text: finalText, tool_calls: toolCalls, turns };
    }

    // Execute every requested tool, feed results back.
    const toolResults = [];
    for (const block of resp.content) {
      if (block.type !== "tool_use") continue;
      const tool = toolMap[block.name];
      let resultText;
      let isError = false;
      try {
        if (!tool) throw new Error(`Unknown tool: ${block.name}`);
        const out = await tool.handler(block.input || {});
        resultText = typeof out === "string" ? out : JSON.stringify(out, null, 2);
      } catch (e) {
        resultText = `Error: ${e.message}`;
        isError = true;
      }
      toolCalls.push({ tool: block.name, input: block.input, ok: !isError });
      toolResults.push({
        type: "tool_result",
        tool_use_id: block.id,
        content: resultText,
        is_error: isError,
      });
    }
    messages.push({ role: "user", content: toolResults });
  }

  return {
    agent: agent.name,
    final_text: `Stopped after hitting the ${MAX_TURNS}-turn limit before finishing.`,
    tool_calls: toolCalls,
    turns,
  };
}
